### DISCUSSION TITLE
How to swap DMA buffers when using double-buffering with RTIC?

### DISCUSSION THREAD

**Discussion by atoktoto** [stm32-rs/stm32f4xx-hal#701] [Q&A] [answered]
I'm feeding i2s through DMA. I ended up with somewhat weird code in my `#[init]`:
```rust
let mut i2s_driver = I2sDriver::new(i2s_peripheral, i2s_driver_config);
i2s_driver.set_tx_dma(true);
i2s_driver.enable();

let buffer1 = cortex_m::singleton!(: [u16; ARRAY_SIZE] = [1; ARRAY_SIZE]).unwrap();
let buffer2 = cortex_m::singleton!(: [u16; ARRAY_SIZE] = [1; ARRAY_SIZE]).unwrap();
let buffer3 = cortex_m::singleton!(: [u16; ARRAY_SIZE] = [1; ARRAY_SIZE]).unwrap(); // but why :D ?

let mut transfer = Transfer::init_memory_to_peripheral(
    stream,
    i2s_driver,
    buffer1,
    Some(buffer3),
    config::DmaConfig::default()
        .memory_increment(true)
        .fifo_enable(true)
        .double_buffer(true)
        .fifo_error_interrupt(true)
        .half_transfer_interrupt(true),
);
transfer.start(|_tx| {});

unsafe {
    cortex_m::peripheral::NVIC::unmask(Interrupt::DMA2_STREAM2);
}

(
    Shared {},
    Local {
        i2s_transfer: transfer,
        buffer: Some(buffer2)
    },
    init::Monotonics(),
)
```

Do I really need 3 buffers or is there a smarter way of dealing with this?

**Comment by burrbull** [comment, marked as answer]
You are right. This part of API looks bad designed.
If you have ideas how to fix this, I'd look at them.

**Comment by atoktoto** [comment]
No idea, I'm just starting with this stuff. In fact it's my first time using DMA directly and I'm still super-confused. Based on C++ tutorials, I though that I should swap buffers on half_transfer and on transfer_complete. But an attempt to do so on half_transfer results in `NotReady` error. Is there something that I'm missing?

**Reply by burrbull** [reply]
It should swap buffers internally, but I'm not sure.

**Reply by atoktoto** [reply]
Looks like it does in fact rotate buffers internally (all 3 of them :D ). I initialized each one to a different value and thats what the output looks like:
![obraz](https://github.com/stm32-rs/stm32f4xx-hal/assets/34739613/ae4828b5-f3ea-4e0f-87ca-32f9c732d1ac)

```rust
if flags.transfer_complete {
    let next_buffer = ctx.local.buffer.take().unwrap();
    let r = transfer.next_transfer(next_buffer);
    let (prev_buffer, _) = r.unwrap();
    *ctx.local.buffer = Some(prev_buffer); 
} 
```
